File: Algorithms/Backtracking/Permutations.java

public class Permutations {

    private static void swap(char[] chars, int a, int b) {

        char temp = chars[a];

        chars[a] = chars[b];
        chars[b] = temp;

    }

    public static void printPermutations(char[] chars, int index) {

        if (index == chars.length - 1) {

            System.out.println(String.valueOf(chars));

        } else {

            for (int i = index; i < chars.length; i++) {

                swap(chars, index, i);

                printPermutations(chars, index + 1);

                swap(chars, index, i);

            }

        }

    }

    public static void main(String[] args) {

        printPermutations("asdf".toCharArray(), 0);

    }

}

File: Algorithms/Backtracking/Permutations.js
"use strict";

let printPermutations = (chars, index = 0) => {

    if (index == chars.length - 1) {

        console.log(chars.join(""));

    } else {

        for (let i = index; i < chars.length; i++) {

            [chars[index], chars[i]] = [chars[i], chars[index]];

            printPermutations(chars, index + 1);

            [chars[index], chars[i]] = [chars[i], chars[index]];

        }

    }

}

printPermutations("asdf".split(''));
